File: Leetcode/C++/98. Validate Binary Search Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return isValid(root, pow(-2, 31)-1, pow(2, 31));
    }
    
    bool isValid(TreeNode* r, long mn, long mx) {
        if(!r) return true;
        if(r->val <= mn || r->val >= mx) return false;
        // cout<<r->val<<endl;
        
        return isValid(r->left, mn, r->val) && isValid(r->right, r->val, mx);
    }
};

File: Leetcode/TypeScript/98. Validate Binary Search Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBST(root: TreeNode | null): boolean {
    return validate(root, Math.pow(-2,31)-1, Math.pow(2,31));
};

const validate = (root: TreeNode | null, l: number, r: number): boolean => {
    if(!root) return true;
    
    if(root.val<=l || root.val>=r) return false;
    
    return validate(root.left, l, root.val) && validate(root.right, root.val, r);
}
